components/eventing-controller/controllers/subscription/nats/integration_test.go

```diff
@@ -0,0 +1,837 @@
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+const (
+	smallTimeout           = 10 * time.Second
+	smallPollingInterval   = 1 * time.Second
+	subscriptionNameFormat = "nats-sub-%d"
+)
+
+const (
+	useExistingCluster       = false
+	attachControlPlaneOutput = false
+)
+
+type testEnsemble struct {
+	testID                    int
+	cfg                       *rest.Config
+	k8sClient                 client.Client
+	testEnv                   *envtest.Environment
+	natsServer                *natsserver.Server
+	reconciler                *natsreconciler.Reconciler
+	natsBackend               *handlers.Nats
+	defaultSubscriptionConfig env.DefaultSubscriptionConfig
+	subscriberSvc             *v1.Service
+	cancel                    context.CancelFunc
+}
+
+type expect struct {
+	k8sSubscription  []gomegatypes.GomegaMatcher
+	k8sEvents        []v1.Event
+	natsSubscription []gomegatypes.GomegaMatcher
+}
+
+//TestUnavailableNATSServer tests if a subscription is reconciled properly when the NATS backend is unavailable.
+func TestUnavailableNATSServer(t *testing.T) {
+	natsPort := 4220
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, natsPort, 7070)
+	defer ens.cancel()
+
+	subscription, subscriptionName := createSubscription(ctx, g, ens,
+		reconcilertesting.WithFilter("", uncleanEventType("")),
+		reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+	)
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName,
+		reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+		reconcilertesting.HaveSubscriptionReady(),
+		reconcilertesting.HaveCleanEventTypes([]string{cleanEventType("")}),
+	)
+
+	ens.natsServer.Shutdown()
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, reconcilertesting.HaveSubscriptionNotReady())
+
+	ens.natsServer = startNATS(natsPort)
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, reconcilertesting.HaveSubscriptionReady())
+}
+
+//TestCreateSubscription tests if subscriptions get created properly.
+func TestCreateSubscription(t *testing.T) {
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, 4225, 7075)
+	defer ens.cancel()
+
+	var testCases = []struct {
+		name               string
+		subscriptionOpts   []reconcilertesting.SubscriptionOpt
+		expect             expect
+		shouldTestDeletion bool
+	}{
+		{
+			name: "create and delete",
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithFilter(reconcilertesting.EventSource, reconcilertesting.OrderCreatedEventTypeNotClean),
+				reconcilertesting.WithWebhookForNATS(),
+				reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+			},
+			expect: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
```

Comment:
nit: While this is completely true, what is important here is that the condition is ready. I wonder if we should put more emphasis on this fact and rename it to `DefaultReadyCondition ` ? 🤔 
What is your opinion on this ?

```suggestion
					reconcilertesting.HaveCondition(reconcilertesting.DefaultReadyCondition("")),
```

What is a bit confusing to me is that a DefaultCondition does not have a default `message` (`""` in this case).